### PR TITLE
Add --node-compat to Pages publish

### LINKED ISSUE #1074: Feature Request: Support --node-compat in wrangler pages build
### What version of `Wrangler` are you using?

2.0.6 or beta

### What operating system are you using?

Mac

### Describe the Bug

App was rewrited from CRA to Remix and i've moved all config from Remix Cloudflare Pages template where it works to my project where i get that error
I've tried to use versions: `beta` and `2.0.6` doesn't work on both of them.

commands that i've used:
```
	"dev": "remix build && run-p dev:*",
    "dev:remix": "remix watch",
    "dev:wrangler": "cross-env NODE_ENV=development wrangler pages dev ./public",
```

### Error:
```
[pages:err] ReferenceError: Buffer is not defined
    at /Repos/app-partners-remix/functions/[[path]].js:10:47417
    at /Repos/app-partners-remix/functions/[[path]].js:1:822
    at /Repos/app-partners-remix/functions/[[path]].js:10:67472
    at /Repos/app-partners-remix/functions/[[path]].js:1:822
    at /Repos/app-partners-remix/functions/[[path]].js:10:103675
    at /Repos/app-partners-remix/functions/[[path]].js:1:822
    at /Repos/app-partners-remix/functions/[[path]].js:10:106256
    at /Repos/app-partners-remix/functions/[[path]].js:1:822
    at /Repos/app-partners-remix/functions/[[path]].js:329:99451
    at SourceTextModule.evaluate (node:internal/vm/module:224:23)

✘ [ERROR] Could not start Miniflare.

error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
ERROR: "dev:wrangler" exited with 1.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

## Edit: I found out package `@emotion/server/create-instance` is responsible for that bug, without it wrangler works.

### PR DESCRIPTION
**Pull request by WalshyDev**
What this PR solves / how to test:

This adds `--node-compat` support to Pages, this is a common feature request and will allow for the usage of Stripe and other such libs within Functions.

Associated docs issues/PR:
N/A

Author has included the following, where applicable:

- [x] Tests
- [x] Changeset

Reviewer has performed the following, where applicable:

- [ ] Checked for inclusion of relevant tests
- [ ] Checked for inclusion of a relevant changeset
- [ ] Checked for creation of associated docs updates
- [ ] Manually pulled down the changes and spot-tested

Fixes #1074

### CODE REVIEW

**Comment by GregBrimble** [comment]
https://github.com/cloudflare/wrangler2/pull/2540#issuecomment-1381121155

**Comment by WalshyDev** [comment]
> I'd like to hold on this until we see where this new nodejs_compat compatibility flag goes, especially since we don't have an in-dash option for this at the moment. I think we're going to quite quickly be at a point where using this polyfill is more harmful than good.

The nodejs compat flag won't have parity for quite a while. In-dash option is also not a blocker, we've done this many times and it takes no time at all to this to the UI. 
Considering we already support this in dev and build, I see no reason to not just add it here. People have wanted it ever since Functions was made. This brings us closer in parity with Workers and for basically 0 cost.

This also means the migration to the new flag is also the same rather than being _another_ thing different.

File: packages/wrangler/src/pages/publish.tsx [hunk 1]
```diff
@@ -300,6 +306,12 @@ export const Handler = async ({
 	let filepathRoutingConfig: string | undefined;
 
 	if (!_workerJS && existsSync(functionsDirectory)) {
+		if (nodeCompat) {
+			console.warn(
+				"Enabling node.js compatibility mode for builtins and globals. This is experimental and has serious tradeoffs. Please see https://github.com/ionic-team/rollup-plugin-node-polyfills/ for more details."
+			);
+		}
```
Comment (JacobMGEvans) [inline packages/wrangler/src/pages/publish.tsx:313]: Could you add a test that runs `wrangler pages publish --node-compat`

**Review by JacobMGEvans — CHANGES_REQUESTED** [review]
Looks great, would like to see an integration test for this before it goes in, probably should go here: https://github.com/cloudflare/wrangler2/blob/main/packages/wrangler/src/__tests__/pages.test.ts 😃

**Comment by JacobMGEvans** [comment]
> > I'd like to hold on this until we see where this new nodejs_compat compatibility flag goes, especially since we don't have an in-dash option for this at the moment. I think we're going to quite quickly be at a point where using this polyfill is more harmful than good.
> 
> The nodejs compat flag won't have parity for quite a while. In-dash option is also not a blocker, we've done this many times and it takes no time at all to this to the UI. Considering we already support this in dev and build, I see no reason to not just add it here. People have wanted it ever since Functions was made. This brings us closer in parity with Workers and for basically 0 cost.
> 
> This also means the migration to the new flag is also the same rather than being _another_ thing different.

I agree with this take on the polyfill `--node-compat` the runtime compat flag won't directly interfere with this implementation, they just cannot be used together.

**Comment by notwedtm** [comment]
Any updates on when this might get merged in?

**Comment by jadbox** [comment]
I'm needing to bump this as it's September now without updates. This feature is critical for using PG via tcp balancing on Cloudflare Pages functions.

**Comment by jadbox** [comment]
I'm blocked from going into production with CF pages due to this bug. This prevents me from using PG tcp pooling in my Page workers.

The other issue is that I'm unable to test against remote data with Pages workers due to this bug:
https://github.com/cloudflare/workers-sdk/pull/2541

**Review by GregBrimble — DISMISSED** [review]
Blocking until we've had internal product roadmap discussions.

**Comment by GregBrimble** [comment]
Rather than accepting this as a arg on deploy, should we instead look at the config of the project in the dashboard and use that so it's all kept in one place? Sort of like how we did D1 when it was in beta and needed the shims?

**Comment by WalshyDev** [comment]
It's already an arg for Workers, if we did move it we'd also need to change dev behaviour to pull from the project which just diverges us more from Workers.

**Comment by GregBrimble** [comment]
But we don't yet have it for many other similar props like compat date and flags. Feels like we should solve syncing config separately to adding this node compat feature (so should stay consistent until we've got that bit figured out in a separate effort).

**Comment by janpio** [comment]
I can confirm that using the [prerelease version mentioned above](https://github.com/cloudflare/workers-sdk/pull/2541#issuecomment-1381123337) indeed let's me _deploy_ a project using `pg` to GitHub Pages that I could not deploy before, as the build was running into [weird errors messages](https://discord.com/channels/595317990191398933/1183055574225670314) (which means that `nodejs_compat` just does not work here). Unfortunately the deployment does not seem to work though, requests get a 522 and the deployment shows "Failed" in the web UI.

@WalshyDev Is this supposed to work on the deployed side as well or is this PR here only one side of the coin and there would need to be a change on the platform as well?

We would really appreciate if this could be added to the stable version of `wrangler`, even if undocumented similar to `--node-compat` for `dev`. Then it could still be removed or updated later when you figure out the long term strategy around node compatiblity.

**Comment by Hebilicious** [comment]
Hi @janpio!
If I understand correctly, you were you able to deploy a project to cloudflare pages with `pg` has a dependency, and then you can use postgres from cloudflare pages without issues?

**Review by petebacondarwin — APPROVED** [review]
This feels like a reasonable addition, given that it is already supported by Pages `dev` and `build` commands.

What's left to get this ready to be merged?

**Comment by janpio** [comment]
@Hebilicious Yes, this change to the CLI unblock usage of `pg` in Cloudflare Pages functions mostly. (It does not work when you also want to use Next.js, but that is a separate problem: https://github.com/cloudflare/next-on-pages/issues/605)

**Comment by petebacondarwin** [comment]
I think a workaround for now is to use `wrangler pages build functions --node-compat build-output-directory=public/_worker.js` to generate the Pages Worker and then upload that using `wrangler publish public`.

Does that work for your use case?

**Comment by admah** [comment]
@WalshyDev it looks like this failed on a formatting issue. Could you fix that and we should be good to merge?

**Review by GregBrimble — CHANGES_REQUESTED** [review]
I think we still first need to discuss and make a clear decision on how this impacts the source of truth for this currently dash-only config setting. This, although aligns with Workers in some ways by exposing this CLI arg in the `deploy` command, misaligns with Pages in others, in that this setting can now be configured completely independently of our source-of-truth: the dashboard.

**Comment by janpio** [comment]
@GregBrimble Can `node_compat` (or `--node-compat`) actually be configured via the UI/dashboard? 

In my (external) understanding `node_compat` is a _build-time_ flag that influences how the CLI builds the code of your worker (by adding a bunch of polyfills that makes for example `pg` work), and `nodejs_compat` is a _runtime_ compatibility flag that influences how the code that has been uploaded is being run on the worker, or what code is available when it runs (like some node APIs in the case of `nodejs_compat`).

**Comment by petebacondarwin** [comment]
Since this PR is very unlikely to land as-is (e.g. we actually want to make nodejs compat flag the approach to take in the long run) and this PR has been sitting around for quite some time, we are going to close it for now. If and when we decide to implement the `node-compat` build time setting for Pages we can reopen or reimplement this.

**Comment by janpio** [comment]
What is the recommended workaround in the meantime @petebacondarwin? Right now it seems that `pg` is not usable with Cloudflare Pages then (as it requires `node-compat`).

**Comment by petebacondarwin** [comment]
@janpio - can you try the workaround here?

https://github.com/cloudflare/workers-sdk/pull/2541#issuecomment-1885006209

**Comment by janpio** [comment]
Oh sorry, we tried that but seem to have forgotten to report back:

1. The command there is invalid, I assume the correct order would be `wrangler pages functions build --node-compat build-output-directory=public/_worker.js`
2. Then it fails during the `publish`:
	```
	▲ [WARNING] `wrangler publish` is deprecated and will be removed in the next major version.
	
	  Please use `wrangler deploy` instead, which accepts exactly the same arguments.
	
	
	
	✘ [ERROR] A compatibility_date is required when publishing. Add the following to your wrangler.toml file:.
	
	      ```
	      compatibility_date = "2024-02-01"
	      ```
	      Or you could pass it in your terminal as `--compatibility-date 2024-02-01`
	  See https://developers.cloudflare.com/workers/platform/compatibility-dates for more information.
	```

Full setup here: https://github.com/prisma/ecosystem-tests/pull/4504 + https://github.com/prisma/ecosystem-tests/actions/runs/7743032416/job/21113520833?pr=4504#step:8:147

**Comment by petebacondarwin** [comment]
OK let me reproduce. The error message seems to imply you are only missing `--compatibility-date` flag?

**Comment by janpio** [comment]
Indeed. Although I now realized that `wrangler publish` publishes a Worker and not Pages. What is required to build the whole Pages "combination" and deploy it?

**Comment by petebacondarwin** [comment]
Yeah sorry about this. It is all a bit messy right now. Trying to work out if there is a magical combination of commands that we could use.

**Comment by petebacondarwin** [comment]
Give this a go: https://github.com/prisma/ecosystem-tests/pull/4574

**Comment by millsp** [comment]
Coming here to report that Pete's workaround is [working](https://github.com/prisma/ecosystem-tests/blob/dev/driver-adapters-wasm/pg-cfpages-basic/run.sh#L15-L19) well :rocket:. Thanks!

**Comment by Hebilicious** [comment]
@petebacondarwin Is there any chance that the workaround could be documented in the Pages docs, or at least here ?
I believe using `pg` in cloudflare pages is quite common.

**Comment by delay** [comment]
Maybe you should consider updating your documentation because as far as I can tell this still isn't working with pg.  The workaround I have found is to use the postgres npm package instead.  https://developers.cloudflare.com/workers/tutorials/postgres/

**Comment by petebacondarwin** [comment]
> Maybe you should consider updating your [documentation ](https://developers.cloudflare.com/workers/tutorials/postgres/) because as far as I can tell this still isn't working with pg and Pages. The workaround I have found is to use the postgres npm package instead of pg.

We should update that to make it clear that the tutorial is for Workers not Pages.

**Pull request closed without merging**